### PR TITLE
Vicariousdrama patch 3 (WIP)

### LINKED ISSUE #9: Unbound Variable
I'm trying to replace my legacy System Overview (the old raspibolt startup script) with Raspibolt Pulse & I'm having a problem getting it started. I followed the instructions in this github, did chmod +x after copying it over, and when I run the script I get the following output:
```

RaspiBolt 3: Sovereign ₿itcoin full node
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
/usr/local/bin/raspibolt: line 413: lnd_color: unbound variable
```
I'm not sure what's going on with this. After I ran it for the first time I checked the code & the only thing I saw that I needed to change was `ext_storage2nd="/data"` as I run my OS on an SD card & keep data on SSD & `/data` is where it's mounted. I changed that & it did not resolve the problem. It looks like line 413 has this in it:
`" \`

the lnd_color variable only shows up on line 397

Any Ideas?

### LINKED ISSUE #11: Throws buckets of errors when running on system startup
Would there be a way to delay the first run of raspibolt-pulse until after all essential components have started?

### LINKED ISSUE #15: Support for CLN and Fulcrum
Would love to be able to have the ability to see CLN as opposed to LND and Fulcrum as opposed to Electrs on startup!

### LINKED ISSUE #18: Failed to read CPU temperature
I have this errror on loading of script:

**cat: /sys/class/thermal/thermal_zone0/temp: No such file or directory**

Screenshot at 2022-11-23 22-24-17

the host is a debian with AMD cpu, this is "sensors" output:

k10temp-pci-00c3
Adapter: PCI adapter
temp1:        +33.4 C  (high = +70.0 C)
                       (crit = +100.0 C, hyst = +99.0 C)

is possible to cat the output of sensors or there is a better way to catch value of temp?

### LINKED ISSUE #19: Dynamically determine storage and 2nd drive mount points
## Background

Depending on if the script is run on the following setups, what gets reported as the storage and 2nd drive may or may not be useful.
* OS on MicroSD, Data on attached SSD - storage will report space free on microsd, and 2nd drive report space free on the SSD
* OS + Data on attached SSD bootable - storage and 2nd drive will report space free on SSD
* OS + Data on attached SSD bootable, MicroSD used as a backup - storage and 2nd drive will report space free on SSD

This is because in the script we make assumptions about what the storage and 2nd drive are.  Specifically, storage is the $/ or root of the file system, and the 2nd drive is either that which is mounted to /mnt/ext (Raspibolt < v3) or /data (Raspibolt >= v3)

## Considerations

The real estate in the rapsibolt-pulse / system overview script is constrained because in part a desire to be able to show all the data within one screen, and stick to no wider then terminal width (80 characters).  Two lines are dedicated to resources, giving us the space for reporting on 2 drives.

### 1. Get drive mounts

We could make enhancements to the script to try to dynamically get drive and mount information.

For example, leverage lsblk to list out mountpoints, and sort to push the root to the top (excluding the boot partition)
```sh
$ lsblk --output MOUNTPOINT | grep / | grep -v /boot | sort
```

Example output on a Raspibolt 2 system using microsd as OS and attached SSD for data:
```sh
/
/mnt/ext
```

A Raspibolt 3 system would appear as follows with no additional drive
```sh
/
```

A Raspibolt 3 system where microsd is used as OS and attached drive for the data
```sh
/
/data
```

And for those using SSD drive for the root, and microsd for backups
```sh
/
/backups
```

### 2. Determine free space based on drive mounts

From the above, once we've determined what to check free space on, that can be done via the existing checks by updating the variables referenced.  The free space check on primary drive doesn't change, as still gets disk free on root.  If the result of the lsblk command above returns more then 1 line (pipe it to `| wc -l`), then the second line can be assigned to the variable `ext_storage2nd`.

Sample code
```sh
drivecount=$(lsblk --output MOUNTPOINT | grep / | grep -v /boot | sort | wc -l)
if [ $drivecount -gt 1 ]; then
  ext_storage2nd=$(lsblk --output MOUNTPOINT | grep / | grep -v /boot | sort | sed -n 2p)
else
  ext_storage2nd=""
fi
```

### LINKED ISSUE #20: Detect appropriate servicename and username for checked services
Over time, some services that can optionally be installed have been done so with different usernames to run them, as well as service names for the autostart on boot systemd scripts.

For Ride the Lightning
- historical:  username of `ridethelightning` and servicename of `ridethelightning`
- current: username of `rtl` and servicename of `rtl`

For Core Lightning / c-lightning
- historical: username of `lightning` and servicename of `lightningd`
- other: username of `cln` and servicename of `cln`
- current: username of `lightningd` and servicename of `lightningd`

Propose the following changes

1. Default the following to match the current values set in the guide

```sh
sn_cln="lightningd"
un_cln="lightningd"
sn_rtl="rtl"
un_rtl="rtl"
```

This will normalize for people installing based on the current guide, but wont take into account disparities to the values over time which may affect past installations.  

2. Enhance the script when checking if services are enabled and active to fall back to past known values, with preference for newest to oldest known

### PR DESCRIPTION
**Pull request by vicariousdrama**
- Git check both LND and CLN, and add Fulcrum and Thunderhub
- Improve usage guidance
- Integrate --mock into the section to get Lightning data
- Convert lightning output to data lines for greater flexibility
- Check enabled, then active for services
- Lightning services try --mock, then LND, then CLN 
- Electrum Servers try: Electrs, then Fulcrum
- Block Explorer Servers try: BTC RPC Explorer
- Lightning Web App try: Ride the Lightning, then Thunderhub
- Only one of each service type is reported on, first enabled.
- Improve error suppression where identified

Fixes #9
Fixes #11
Fixes #15
Fixes #18
Fixes #19
Fixes #20

### CODE REVIEW

**Comment by vicariousdrama** [comment]
Tested against 
- Node 1: Bitcoin Core v23.0.0, LND (0.15.4-beta), Ride the Lightning (v0.12.3)  [live funds]
- Node 2: Bitcoin Core v23.0.0, Fulcrum (1.8.2) and Core Lightning (v22.11rc2)  [unfunded]

**Comment by vicariousdrama** [comment]
Sample output with Fulcrum and Lightning.  This is from a windows command prompt using cygwin ssh client so there may be some disparities in formatting for unicode characters that can disrupt the flow of width for  the labels in the lightning section.
image

**Review by blckbx — COMMENTED** [review]
Code changes look good to me

**Review by M1ch43lV — APPROVED** [review]
Well done! Running the scripts on a VM based Ubuntu linux system worked flawlessly except for temperature read-out, because my raspibolt system is not a Pi!

**Review by blckbx — APPROVED** [review]
Approved after partial tests by different users.

File: raspibolt-pulse-switch.sh [hunk 1]
```diff
@@ -47,7 +68,7 @@ trap trap_ctrlC SIGINT SIGTERM
 # print usage information for script
 usage() {
   echo "RaspiBolt Welcome: system status overview
-usage: bbb-cmd.sh [--help] [--mock]
+usage: $0 [--help] [--mock]
```
Comment (kristapsk) [inline raspibolt-pulse-switch.sh]: nit
```suggestion
usage: $(basename "$0") [--help] [--mock]
```

**Comment by kristapsk** [comment]
With these changes it thinks Electrs and BTC RPC Explorer is down on my node, RTL not installed and LND not up to date.

main:
image

this PR:
image

**Comment by vicariousdrama** [comment]
Good catch on Electrs and the Block Explorer @kristapsk.  Those should be resolved now.

I'm looking into how best to handle the RTL indicating as not installed.  I've got variables at the top of the script to set the service name for it which in the past can be either rtl or ridethelightning.  Current guidance per https://github.com/raspibolt/raspibolt/blame/master/guide/lightning/web-app.md#L208 is to use rtl, so I'm going to change the default to that, as well as open an issue to try to detect this when run

**Comment by kristapsk** [comment]
> I'm looking into how best to handle the RTL indicating as not installed. I've got variables at the top of the script to set the service name for it which in the past can be either rtl or ridethelightning. Current guidance per https://github.com/raspibolt/raspibolt/blame/master/guide/lightning/web-app.md#L208 is to use rtl, so I'm going to change the default to that, as well as open an issue to try to detect this when run

I think that's correct approach.

Alternative would be to decide some kind of separator, like `|`, then set variable to `rtl|ridethelightning`, then add bash function that splits string by `|` and checks each one. But I think that would be overkill, don't see the need to support some old legacy configs. It's easier to just ask people to rename or reinstall stuff.

**Comment by vicariousdrama** [comment]
> Alternative would be to decide some kind of separator, like |, then set variable to rtl|ridethelightning, then add bash function that splits string by | and checks each one. But I think that would be overkill, don't see the need to support some old legacy configs. It's easier to just ask people to rename or reinstall stuff.

Latest 2 commits handle service name fallback for those edge cases, as well as username lookups where needed. I'm hoping this covers it good enough for any disparities.

File: raspibolt-pulse-switch.sh [hunk 2]
```diff
@@ -103,14 +103,13 @@ echo -ne '\r### Loading System data \r'
...
 
-# get CPU temp
-cpu=$(cat /sys/class/thermal/thermal_zone0/temp)
-# cpu=$(cat /sys/class/thermal/thermal_zone3/temp)
-
-temp=$((cpu/1000))
-if [ ${temp} -gt 60 ]; then
+# get highest reported temperature
+hitemp=0
+for i in /sys/class/hwmon/hwmon*/temp*_input; do readtemp=$(cat $i); if (( readtemp > hitemp )); then hitemp=$readtemp; fi; done
```
Comment (kristapsk) [inline raspibolt-pulse-switch.sh]: On Odroid HC4, likely this gives me error:
```
cat: '/sys/class/hwmon/hwmon*/temp*_input': No such file or directory
```
`main` branch doesn't give error and displays CPU temperature correcly.

```
admin ~/raspibolt-pulse ₿ cat /sys/class/thermal/thermal_zone0/temp
48800
admin ~/raspibolt-pulse ₿ ls -al /sys/class/hwmon/hwmon0/
total 0
drwxr-xr-x 3 root root    0 Jan 19 21:47 .
drwxr-xr-x 3 root root    0 Jan  1  1970 ..
lrwxrwxrwx 1 root root    0 Jan 19 21:45 device -> ../../../pwm-fan
-rw-r--r-- 1 root root 4096 Jan 19 21:45 fan_speed
-r--r--r-- 1 root root 4096 Jan 19 21:45 name
lrwxrwxrwx 1 root root    0 Jan 19 21:45 of_node -> ../../../../../firmware/devicetree/base/pwm-fan
drwxr-xr-x 2 root root    0 Jan 19 21:45 power
-rw-r--r-- 1 root root 4096 Jan 19 21:45 pwm1
lrwxrwxrwx 1 root root    0 Jan  1  1970 subsystem -> ../../../../../class/hwmon
-rw-r--r-- 1 root root 4096 Jan  1  1970 uevent
```

**Comment by kristapsk** [comment]
Also, there is still warning about LND v0.15.5-beta not being latest version.

File: raspibolt-pulse-switch.sh (hunk 2 above)
Comment (vicariousdrama) [reply raspibolt-pulse-switch.sh]: argh. so on this platform they don't have temp sensors under hwmon?
I guess we can have fallback to the older thermal zones if there's no files matching the glob pattern

**Comment by vicariousdrama** [comment]
> Also, there is still warning about LND v0.15.5-beta not being latest version.

Can you report what you get for each of these

```
lncli getinfo | jq -r '.version | split("=") | .[1]'
```

```
curl -s --connect-timeout 5 https://api.github.com/repos/lightningnetwork/lnd/releases/latest | jq -r '.tag_name | select(.!=null)'
```

**Comment by kristapsk** [comment]
> > Also, there is still warning about LND v0.15.5-beta not being latest version.
> 
> Can you report what you get for each of these
> 
> ```
> lncli getinfo | jq -r '.version | split("=") | .[1]'
> ```
> 
> ```
> curl -s --connect-timeout 5 https://api.github.com/repos/lightningnetwork/lnd/releases/latest | jq -r '.tag_name | select(.!=null)'
> ```

These have answers you might expect:
```
admin ~/raspibolt-pulse ₿ lncli getinfo | jq -r '.version | split("=") | .[1]'
v0.15.5-beta
admin ~/raspibolt-pulse ₿ curl -s --connect-timeout 5 https://api.github.com/repos/lightningnetwork/lnd/releases/latest | jq -r '.tag_name | select(.!=null)'
v0.15.5-beta
```

**Comment by vicariousdrama** [comment]
> These have answers you might expect:
> 
> ```
> admin ~/raspibolt-pulse ₿ lncli getinfo | jq -r '.version | split("=") | .[1]'
> v0.15.5-beta
> admin ~/raspibolt-pulse ₿ curl -s --connect-timeout 5 https://api.github.com/repos/lightningnetwork/lnd/releases/latest | jq -r '.tag_name | select(.!=null)'
> v0.15.5-beta
> ```

ok, now lets check the file that keeps track of what versions its seen in github

```
cat ~/.raspibolt.versions.json
```

**Comment by kristapsk** [comment]
Everything seems correct there too.
```
admin ~/raspibolt-pulse ₿ cat ~/.raspibolt.versions.json | grep lnd
    "lnd": "v0.15.5-beta",
```

**Comment by vicariousdrama** [comment]
Any chance your get_LND_data.sh may be from an older commit ?

**Comment by vicariousdrama** [comment]
Before updating some out of date services
image


After updating LND and RTL on my node
image


setup: this is raspberry pi 4, 8 GB, and RTL is installed under ridethelightning user and has ridethelightning as service name

to test branch
```
# clone it
cd /tmp
git clone -b vicariousdrama-patch-3 https://github.com/vicariousdrama/raspibolt-pulse.git
cd raspibolt-pulse
chmod +x raspibolt-pulse-switch.sh
# test run before install
./raspibolt-pulse-switch.sh
```

when satisfied, install it
```
sudo cp *.sh /usr/local/bin/
sudo mv /usr/local/bin/raspibolt-pulse-switch.sh /usr/local/bin/raspibolt
# run as raspibolt
cd ~
raspibolt
```

**Comment by brokenthunder** [comment]
2a511f66-b2c9-47fe-9045-7edc0ea3bbde

**Comment by vicariousdrama** [comment]
Sample of running on a system without Bitcoin and related services 
image

**Comment by zynos** [comment]
Can we merge this branch now?

**Review by blckbx — APPROVED** [review]
lgtm

**Pull request merged**